### PR TITLE
Added new ReactARTFiber target based on ReactFiberReconciler

### PR DESCRIPTION
**Pull request by bvaughn**
Also updated existing unit tests to run against both legacy and fiber ReactART builds.

The new target also fixes a bug with event handlers that exists in the legacy target. I believe the existing react-art renderer intended to use [`handleEvent`](https://github.com/reactjs/react-art/blob/db4dcef9532cc9e3e1235bc62e917efcb3fb5b24/src/ReactART.js#L291-L301) as the event subscriber but was [actually using the `props.listener`](https://github.com/reactjs/react-art/blob/db4dcef9532cc9e3e1235bc62e917efcb3fb5b24/src/ReactART.js#L281). This mostly worked, but failed in the case where a specific event handler (eg `onClick`) changed.

While testing, I noticed that the `ClippingRectangle` target doesn't work quite right for SVG mode. This is not a regression. It's something that needs to be fixed in the underlying Art library (if we care).

### CODE REVIEW

File: examples/vector-widget/VectorWidget.js [hunk 1]
```diff
@@ -8,93 +8,72 @@
...
-var Shape = ReactART.Shape;
-var Surface = ReactART.Surface;
-var Transform = ReactART.Transform;
-
-var MOUSE_UP_DRAG = 0.978;
-var MOUSE_DOWN_DRAG = 0.9;
-var MAX_VEL = 11;
-var CLICK_ACCEL = 3;
-var BASE_VEL = 0.15;
+const React = require('react');
```
Comment (sebmarkbage) [inline examples/vector-widget/VectorWidget.js]: Can we split this file out into a separate PR? It's unclear if you needed to change something here to support the new renderer or if it is just a style change. Regardless it's unrelated.

File: src/ReactART.js [hunk 2]
```diff
@@ -329,315 +88,446 @@ const NodeMixin = {
...
+    this._surface = Mode.Surface(+width, +height, this._tagRef);
+
+    // HACK Simplify appendChild-type methods by mirroring Art instance wrappers
+    const container = {
+      instance: this._surface,
+    };
+
+    this._mountNode = ARTRenderer.mountContainer(
+      this.props.children,
+      container,
```
Comment (sebmarkbage) [inline src/ReactART.js]: This no longer passes context along. I think we'll need to provide the third argument here.

File: src/ReactART.js [hunk 3]
```diff
@@ -329,315 +88,446 @@ const NodeMixin = {
...
-        oldProps.alignment !== props.alignment ||
-        oldProps.path !== props.path) {
-      this.node.draw(
-        newString,
-        props.font,
-        props.alignment,
-        props.path
-      );
-      this._oldString = newString;
+    ARTRenderer.updateContainer(this.props.children, this._mountNode);
```
Comment (sebmarkbage) [inline src/ReactART.js]: This also needs to pass the third argument.

File: src/ReactART.js [hunk 4]
```diff
@@ -329,315 +88,446 @@ const NodeMixin = {
...
+        break;
+      case TYPES.TEXT:
+        instance = new Text(props);
+        break;
+    }
+
+    return instance;
+  },
+
+  createTextInstance(text, internalInstanceHandle) {
```
Comment (sebmarkbage) [inline src/ReactART.js]: I believe that Fiber will turn any array of strings into `createTextInstance` calls. So I think that something like `<Text>{"foo"}{"bar"}</Text>` will break.

File: src/ReactART.js [hunk 5]
```diff
@@ -12,312 +12,71 @@
...
-    node.inject(parentNode);
+    return '';
   }
 }
 
-// ContainerMixin for components that can hold ART nodes
+/** COMPONENTS */
 
-const ContainerMixin = assign({}, ReactMultiChild, {
+class Node {
```
Comment (sebmarkbage) [inline src/ReactART.js]: You have a wrapper class around every instance. However, they don't seem to have much interesting state associated with them so we're allocating them only to store a pointer to the ART instance.

Could you just use the ART instance as the instance and rely on pattern matching to pick the logic branch instead of OO methods?

File: src/ReactART.js [hunk 6]
```diff
@@ -12,312 +12,71 @@
...
-   * @param {ReactComponent} child Component to move.
-   * @param {number} toIndex Destination index of the element.
-   * @protected
-   */
-  moveChild: function(child, afterNode, toIndex, lastIndex) {
-    const childNode = child._mountImage;
-    injectAfter(this.node, afterNode, childNode);
-  },
+    this._listeners = {};
+    this._subscriptions = {};
```
Comment (sebmarkbage) [inline src/ReactART.js]: We don't need to store subscriptions because we can just rely on garbage collection to do the clean up for us. In fact, `_removeEventListeners` isn't even called.

I suppose the current ART API requires us to store the subscriptions to unsubscribe when you swap them out during updates though. Maybe we can just store them as an expando property on the ART instance as a hack?

In the normal case we don't actually use many of these listeners anyway. We attach top level listeners at the root and dispatch it ourselves.

File: examples/vector-widget/VectorWidget.js (hunk 1 above)
Comment (bvaughn) [inline examples/vector-widget/VectorWidget.js]: Sure. This file was causing some console warnings before:

> Warning: bind(): React component methods may only be bound to the component instance. See VectorWidget

No changes were _required_ though. I just fixed the warnings and added a little better react-art component coverage. I can do both via a separate PR.

File: src/ReactART.js (hunk 4 above)
Comment (bvaughn) [inline src/ReactART.js]: Ah. You are correct.

I'm unsure of how to best handle this since the `text` param is just a string with no associated props (eg font, color). At a glance, I _think_ this information is accessible via `internalInstanceHandle.return.pendingProps` but I'm not sure if that would be acceptable or hacky, eg:

```js
  createTextInstance(text, internalInstanceHandle) {
    const props = Object.assign(
      {},
      internalInstanceHandle.return.pendingProps,
      { text }
    );

    return new Text(props);
  },
```

File: src/ReactART.js (hunk 2 above)
Comment (bvaughn) [inline src/ReactART.js]: This may be a stupid question...but...why do ReactART components need `context`? The set of components being managed by the `ARTRenderer` only includes the ones explicitly defined in this module, none of which use `context`.

File: src/ReactART.js (hunk 4 above)
Comment (spicyj) [inline src/ReactART.js]: Hm, one option could be to make Text a composite that flattens the children and renders the lower-level host text component.

Comment (spicyj) [inline src/ReactART.js]: Seb probably has other ideas.

File: src/ReactART.js (hunk 2 above)
Comment (bvaughn) [inline src/ReactART.js]: Gotcha. That makes sense. Thanks!

File: src/ReactART.js (hunk 4 above)
Comment (bvaughn) [inline src/ReactART.js]: Yeah, I think I'll go with something like that after trying this as a first pass.

Comment (bvaughn) [inline src/ReactART.js]: This actually turned out to be easier to handle than I realized at first. `createTextInstance` has to return something non-null (eg the `text` string itself). `appendInitialChild` has to be smart enough to handle the case where `child` is that string (eg a no-op case). And then `Text` itself handles the fact that it may have a string-child or array-of-string-children.

File: src/ReactART.js [hunk 7]
```diff
@@ -329,315 +92,466 @@ const NodeMixin = {
...
+        instance = new ClippingRectangle(props);
+        break;
+      case TYPES.GROUP:
+        instance = new Group(props);
+        break;
+      case TYPES.SHAPE:
+        instance = new Shape(props);
+        break;
+      case TYPES.TEXT:
+        instance = new Text(props);
```
Comment (edvinerikson) [inline src/ReactART.js]: does Text handle props.children as well? since `<Text>Hello</Text>` will not create a text instance.

Comment (bvaughn) [inline src/ReactART.js]: Yes. `Text` handles single-children (eg `<Text>Hello</Text>`) and arrays of children (eg `<Text>Hello, {name}</Text>`).x

File: src/ReactART.js (hunk 5 above)
Comment (bvaughn) [inline src/ReactART.js]: Totally. I've removed the wrapper objects. The metadata I was storing on the wrapper instance (eg current `Text` string, previous path delta, etc) is now just being stored on the ART-created instance. Kind of hacky but it avoids the wrapper object.

File: src/ReactART.js (hunk 4 above)
Comment (spicyj) [inline src/ReactART.js]: I guess that by reading props directly, we sacrifice being able to include components that return text:

```
const CurrentTime = () => String(new Date());

<Text>It's now <CurrentTime />.</Text>
```

This doesn't work in the old version so not a regression but ideally we could support this. Probably doesn't matter for ART but we'll want this to work for RN.

Comment (bvaughn) [inline src/ReactART.js]: Good point Ben! Let me think about that a bit more. 😄

Comment (bvaughn) [inline src/ReactART.js]: I think I may punt on supporting custom components that return text for this initial iteration. We can always add support for that in a follow-up.

The current API makes it difficult to associate a text string with its parent instance. I initially tried:

1. `createTextInstance` returns `text` (string).
1. `appendInitialChild` handles special case where `child` is a string by appending to an array stored on the ART text `instance` (for lack of a better place).
1. When applying props for text, use the aforementioned array whenever `props.children` is not a string.

Unfortunately `commitTextUpdate` only gets the updated text string and has no reference to the ART instance, so no way to update the text-array.

Also, `createTextInstance` is called _before_ the ART instance is created and so the only thing I have access to at that point is the `internalInstanceHandle` which I shouldn't be crawling in the renderer.  This also complicates the _initial_ display because- without using `internalInstanceHandle`- there's no way to gather up text-children and pass them along to the eventual ART text instance.

**Comment by bvaughn** [comment]
> We can't really land this as is to replace ReactART. Maybe we should land something as ReactARTFiber first so that we can test it out first?

Sure! I added a new `ReactARTFiber.js` file. To test it locally, I'm just updating `package.json` so that it's main entry points at the fiber version (eg `"main": "lib/ReactARTFiber.js",`). Not sure how best to bundle this up for local, opt-in testing. Suggestions welcome. 😄

File: src/ReactART.js (hunk 4 above)
Comment (sebmarkbage) [inline src/ReactART.js]: You would need to have a place holder object for text instances with a parent pointer and store an expando with the list of text instances on the Text ART instance. So that when commitTextUpdate happens, you can do `textInstance.parent.draw(textInstance.parent._textInstances.map(ti => ti.text).join(''))`

Comment (sebmarkbage) [inline src/ReactART.js]: Maybe we need an alternative text API that makes it easier for certain renderers. I'm not sure if it can be much more efficient than this though since these can be nested deeply. So maybe a helper lib on top of host config would be enough.

Comment (bvaughn) [inline src/ReactART.js]: I tried that as well, but it didn't work for reasons I explained above. (Maybe one of us is misunderstanding each other?) If it would be easier to chat on Hangouts I could. 😄

Comment (bvaughn) [inline src/ReactART.js]: > You would need to have a place holder object for text instances with a parent pointer 

The problem is that when the text is first created (`createTextInstance`) I only have access to the text string itself and the `internalInstanceHandle`. I _could_ crawl the `return` pointers to find the parent text node- but that's probably not a good idea to do in the renderer.

Also, the ART `Text` instance doesn't exist yet at this time, so I would have to also bundle along the text strings to that instance when I later created it.

Overall it felt dirty/hacky.

File: src/ReactARTFiber.js [hunk 8]
```diff
@@ -0,0 +1,496 @@
...
+require('art/modes/current').setCurrent(
+  // Change to 'art/modes/dom' for easier debugging via SVG
+  require('art/modes/fast-noSideEffects'),
+);
+
+const Mode = require('art/modes/current');
+const Transform = require('art/core/transform');
+const invariant = require('fbjs/lib/invariant');
+const React = require('react');
+const ReactFiberReconciler = require('react-dom/lib/ReactFiberReconciler');
```
Comment (iamdustan) [inline src/ReactARTFiber.js]: curious, are you testing this against React@master?

Comment (bvaughn) [inline src/ReactARTFiber.js]: Yes. Pulled as recently as an hour or two ago. Why?

Comment (bvaughn) [inline src/ReactARTFiber.js]: Oh! Gotcha. 😄

I tried using `npm link` at first but ran into problems with Babel trying to reprocess the dist for 'react' and 'react-dom'. Seems like the symlink created by `npm link` caused the exclude for 'node_modules' not to work for those packages. So I gave up on that after a while.

What I ended up doing was a bit of a hack- but it worked, so...

I checked out React master and copied the built versions of 'react' and 'react-dom' into the `node_modules` folder of 'react-art'. So basically...

```
cd path/to/react
cp -r ./build/packages/react path/to/your/project/node_modules/
cp -r ./build/packages/react-dom path/to/your/project/node_modules/
```

Comment (iamdustan) [inline src/ReactARTFiber.js]: Thanks! :smile:

I wonder if http://react.zpao.com/builds/master/latest/ is still kept up to date for a daily sync, too...

Comment (bvaughn) [inline src/ReactARTFiber.js]: That looks like UMD builds, which won't provide a way to load the new fiber reconciler.

Comment (iamdustan) [inline src/ReactARTFiber.js]: I *think* the react-dom.tgz files are the packages for npm assets...

Comment (spicyj) [inline src/ReactARTFiber.js]: Yes.

**Comment by bvaughn** [comment]
I've updated the unit tests to run against both version of ReactART. In order to get the tests working, I had to replace string refs with function refs. (Otherwise `ReactFiberCommitWork` throws an error in `attachRef` where it [seemingly expects all `ref`s to be functions](https://github.com/facebook/react/blob/df9dc8933387c5f1e6126ce38e95771c890f8e4f/src/renderers/shared/fiber/ReactFiberCommitWork.js#L65-L70). I assume this is a known issue?)

**Comment by spicyj** [comment]
String refs should work; grep for coerceRef.

**Comment by bvaughn** [comment]
For example, this test fails with `TypeError: ref is not a function`:
```jsx
    it('resolves refs before componentDidUpdate', function() {
      var CustomShape = React.createClass({
        render: function() {
          return <Shape />;
        }
      });
      var ref = {};
      var Outer = React.createClass({
        componentDidMount: function() {
          ref = this.refs.test;
        },
        componentDidUpdate: function() {
          ref = this.refs.test;
        },
        render: function() {
          return (
            <Surface>
              <Group>
                {this.props.mountCustomShape && <CustomShape ref="test" />}
              </Group>
            </Surface>
          );
        }
      });
      var container = document.createElement('div');
      ReactDOM.render(<Outer />, container);
      expect(ref).not.toBeDefined();
      ReactDOM.render(<Outer mountCustomShape={true} />, container);
      expect(ref.constructor).toBe(CustomShape);
    });
```

`coerceRef` is failing to wrap the string ref here. It's checking `element.ref` (which is `null`). The ref in this case is actually `element._owner.ref`. I haven't really dug into this, just some really quick inspecting.

**Comment by bvaughn** [comment]
This change set was approved in Phabricator. I'm going to merge here to keep Github in sync.
